### DISCUSSION TITLE
Async, CSRF, headers, cookies

### DISCUSSION THREAD

**Discussion by SkyHyve** [alorence/django-modern-rpc#51] [General]
Hi,

I am looking at this project and might want to give it a go, but I have a few questions:

1. Does the project support async procedures/function just by defining them with the _async_ keyword?
2. Is there CSRF validation in the procedures / how to implement?
3. In terms of headers and cookies, given that the precedures only take params rather than requests/responses how are headers and cookies supposed to be implemented on a procedure by procedure basis?

Thanks

**Comment by alorence** [comment]
Hi Matthew,

Thank you for your interest in the project.

1. There is no async procedures support for now
2. I am not sure CSRF validation process is required to receive RPC requests, but this may be checked manually, see 3.
3. Indeed procedures takes params, but you can add the special `**kwargs` argument to any procedure to get a dict with some useful objects, including the current request instance. See https://django-modern-rpc.readthedocs.io/en/latest/basics/register_procedure.html#access-internal-information Reading the request allows you to check for cookies, headers, CSRF, etc. These kind of checks may be done in a predicate, as part of an authentication procedure. See https://django-modern-rpc.readthedocs.io/en/latest/advanced/authentication.html In such case, the predicate takes only the request as argument and must return a boolean value

**Reply by SkyHyve** [reply]
Awesome thanks for the prompt reply.

Unfortunately I need async support but I don't need RPC immediately. Are there plans for async support in the near future? Not entirely sure of how much of an undertaking it is

**Reply by alorence** [reply]
Implementing async procedures shouldn't be a huge deal, but to be honest I will miss time in the next weeks. This is something I already thought about, but I really don't know when I will have time to work on.

But if you don't need RPC but async support, this library is probably not what you need... It is exclusively dedicated to provide XML-RPC & JSON-RPC endpoints to your Django project.

**Reply by SkyHyve** [reply]
Ah thanks and yeah I understand. However, excuse my ignorance, but why exactly would RPC be mutually exclusive with async? I haven't worked with RPC much to know what the issue is. As far as I can tell its just a specific type of HTTP POST request

**Reply by alorence** [reply]
async and RPC are not mutually exclusive. But you said:

> I need async support but I don't need RPC immediately

Since this library provide RPC features, if you don't need RPC, you probably don't need this library... And since it doesn't provide async support, I am not sure to understand your use case.

**Reply by SkyHyve** [reply]
> async and RPC are not mutually exclusive. But you said:
> 
> > I need async support but I don't need RPC immediately
> 
> Since this library provide RPC features, if you don't need RPC, you probably don't need this library... And since it doesn't provide async support, I am not sure to understand your use case.

Sorry the keyword here is _immediately_ so I will need it at some point in the nearish future and this project looks like the most promising

**Comment by alorence** [comment]
To complete my answer about support of async procedures, the biggest problem right now is the compatibility with python 3.5 and 3.6.

These versions will be dropped in the next major release, 1.1. I will try to introduce async support in this next milestone

**Comment by alorence** [comment]
Hy @SkyHyve,

I know you opened this discussion a long time ago. I wanted to send you some updates about requested features.

The just published v2.0 release introduce the support of full async procedures. Simply declare a procedure as coroutine with `async` keyword and it will work seamlessly. For better performances, use the new `server.async_view` as a replacement of `server.view` and all your procedures will be executed in a full async view.

The is still no support for CSRF validation, as it does not really make sense for XML-RPC / JSON-RPC requests. The views receiving such requests are configured with csrf_exempt() decorator. You can still perform the checks you need using the authentication system.

The new argument `context_target` allows you to define the procedure arg that will receive the RpcRequest object, allowing to access the full HttpRequest object (including raw body, headers, etc.). This replaces the old `**kwargs` trick and provide similar functionalities
See https://django-modern-rpc.readthedocs.io/latest/config/procedure_registration.html#accessing-the-context
